### PR TITLE
User can create a todo.

### PR DESCRIPTION
**Pull request by ryanchangyeolshin**
image

image

### CODE REVIEW

**Comment by thebearingedge** [comment]
Did you mean to send a `.txt` to `occs-ten`?

**Comment by ryanchangyeolshin** [comment]
Not yet. Wanted to double check my work. The tests seem to run fine, but can't be too sure.

File: client/todo-form.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  constructor(props) {
+    super(props)
+    this.handleSubmit = this.handleSubmit.bind(this)
+  }
+
+  handleSubmit(e) {
+    e.preventDefault()
+    const formData = new FormData(e.target)
+    const data = {
+      _id: uuid(),
```
Comment (thebearingedge) [inline client/todo-form.js]: Don't create the `uuid` here, let your server do it.

File: client/todo-form.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      dueDate: formData.get('dueDate'),
+      task: formData.get('task')
+    }
+    axios.post('http://localhost:3000/api/todos', data)
```
Comment (thebearingedge) [inline client/todo-form.js]: Just hit `'/api/todos'`. You don't want to hard-code your `origin`.

File: package.json [hunk 3]
```diff
@@ -32,12 +37,22 @@
   "homepage": "https://github.com/ryanchangyeolshin/continuous-delivery#readme",
   "dependencies": {
     "axios": "^0.17.0",
+    "body-parser": "^1.18.2",
     "dotenv": "^4.0.0",
     "express": "^4.16.2",
     "mongodb": "^2.2.33",
+    "path": "^0.12.7",
```
Comment (thebearingedge) [inline package.json]: `path` is built into Node, you don't need to `npm install` it.

File: server/create-app.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    .use(express.static(publicPath))
+    .use(bodyParser.json())
+    .get('/api', (req, res) => {
+      res.status(200).json({
+        name: 'continuous-delivery',
+        description: 'A practice repository for testing and deployment'
+      })
+    })
+    .get('/api/todos', async (req, res) => {
+      const _todos = await findTodos(todos)
```
Comment (thebearingedge) [inline server/create-app.js]: Doesn't `findTodos` have a closure over the collection?

File: sync.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+require('dotenv/config')
+
+module.exports = {
+  // server: 'server/public',
```
Comment (thebearingedge) [inline sync.js]: You can delete this comment.

File: test/index.test.js [hunk 6]
```diff
@@ -39,20 +39,31 @@ describe('app', () => {
     await _todos.insertOne(_todo)
   })
 
-  describe('GET /', () => {
+  describe('GET /api', () => {
     it('should return a json object with the repository name and description', async () => {
-      const { data } = await axios.get('http://localhost:3000/')
-      expect(data).to.deep.equal(_repo)
+      const repo = await axios.get('http://localhost:3000/api')
```
Comment (thebearingedge) [inline test/index.test.js]: Use your `process.env.PORT` instead of hard-coding it.

File: server/todos-gateway.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+module.exports = collection => {
+  return {
+    async findTodos() {
+      const todos = await collection.find({}).toArray()
+      return todos
+    },
+    async insertTodo(data) {
+      const todo = await collection.insertOne(data)
+      return todo
```
Comment (thebearingedge) [inline server/todos-gateway.js]: Just return the inserted `todo` instead of the MongoDB `result` object.

File: test/todos-gateway.test.js [hunk 8]
```diff
@@ -33,10 +36,18 @@ describe('todos-gateway', () => {
...
+      const todos = await findTodos(_todos)
       expect(todos[0]).to.deep.equal(_todo)
     })
   })
+
+  describe('insertTodo()', () => {
+    it('return the insertedTodo', async () => {
+      const data = { _id: uuid(), dueDate: '1/1/2018', 'task': 'Go for a jog.' }
+      const { ops } = await insertTodo(data)
+      expect(ops[0]).to.deep.equal(data)
```
Comment (thebearingedge) [inline test/todos-gateway.test.js]: Notice how your test has to reach into the `ops`. Just make your function return what you really want.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

**Review by thebearingedge — APPROVED** [review]

**Pull request merged**